lib/Dialects/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -208,5 +208,142 @@ static ParseResult parseTileOp(OpAsmParser &parser, OperationState &result) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// InParallelOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verify(InParallelOp op) {
+  // Check that the body defines as single block argument for the thread index.
+  auto *body = op.getBody();
+  if (body->getNumArguments() != 1)
+    return op.emitOpError("body expects exactly one argument");
+  if (!body->getArgument(0).getType().isIndex())
+    return op.emitOpError(
+        "expected body first argument to be an index argument for "
+        "the thread index");
+
+  // Verify consistency between the result types and the terminator.
+  auto terminatorTypes =
+      llvm::cast<PerformConcurrentlyOp>(body->getTerminator()).resultTypes();
+  auto opResults = op.getResults();
+  if (opResults.size() != terminatorTypes.size())
+    return op.emitOpError("has ")
+           << opResults.size() << " results, but its terminator produces "
+           << terminatorTypes.size() << " values";
+  unsigned i = 0;
+  for (auto e : llvm::zip(terminatorTypes, opResults)) {
+    if (std::get<0>(e) != std::get<1>(e).getType())
+      return op.emitOpError()
+             << "type mismatch between " << i << "th result of in_parallel ("
+             << std::get<0>(e) << ") and " << i << "th result produced by its "
+             << "terminator (" << std::get<1>(e).getType() << ")";
+    i++;
+  }
+
+  return success();
+}
+
+static void print(OpAsmPrinter &p, InParallelOp op) {
+  p << ' ' << op.num_threads() << ' ';
+  p << " -> (" << op.getResultTypes() << ')';
+  p.printRegion(op.region(),
+                /*printEntryBlockArgs=*/true,
+                /*printBlockTerminators=*/true);
+  p.printOptionalAttrDict(op->getAttrs());
+}
+
+static ParseResult parseInParallelOp(OpAsmParser &parser,
+                                     OperationState &result) {
+  auto &builder = parser.getBuilder();
+
+  OpAsmParser::OperandType numThreads;
+  Type indexType = builder.getIndexType();
+
+  if (parser.parseOperand(numThreads) ||
+      parser.resolveOperand(numThreads, indexType, result.operands))
+    return failure();
+  if (parser.parseArrowTypeList(result.types)) return failure();
+
+  SmallVector<OpAsmParser::OperandType, 8> regionOperands;
+  SmallVector<Type, 8> regionTypes;
+  std::unique_ptr<Region> region = std::make_unique<Region>();
+  if (parser.parseRegion(*region, regionOperands, regionTypes))
+    return failure();
+  InParallelOp::ensureTerminator(*region, builder, result.location);
```

Comment:
Somewhere the BB produced by this call is missing the index barg and other things fail downstream.

Please send a fix and a proper test soon after landing this.

Reply:
Yes, that's expected. `ensureTerminator` only ensures that the block exists and is terminated, but it's not responsible for adding any necessary args. For example, the implementation of `ensureTerminator` for `scf.for` doesn't insert bbargs for the induction variable and its `iter_args`.